py/server/deephaven/experimental/table_data_service.py

```diff
@@ -0,0 +1,339 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+"""This module defines a table service backend interface that users can implement to provide external data to Deephaven
+tables."""
+
+from abc import ABC, abstractmethod
+from typing import Tuple, Optional, Callable
+
+import jpy
+
+import pyarrow as pa
+
+from deephaven.dherror import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.table import Table
+
+_JPythonTableDataService = jpy.get_type("io.deephaven.extensions.barrage.util.PythonTableDataService")
+_JTableKeyImpl = jpy.get_type("io.deephaven.extensions.barrage.util.PythonTableDataService$TableKeyImpl")
+_JTableLocationKeyImpl = jpy.get_type("io.deephaven.extensions.barrage.util.PythonTableDataService$TableLocationKeyImpl")
+
+
+class TableKey(ABC):
+    """A key that identifies a table. The key should be unique for each table. The key can be any Python object and
+    should include sufficient information to uniquely identify the table for the backend service."""
+
+    @abstractmethod
+    def __hash__(self):
+        pass
+
+
+class TableLocationKey(ABC):
+    """A key that identifies a specific location of a table. The key should be unique for each table location of the
+    table. The key can be any Python object and should include sufficient information to uniquely identify the location
+    for the backend service to fetch the data values and data size.
+    """
+
+    @abstractmethod
+    def __hash__(self):
+        pass
+
+
+class TableDataServiceBackend(ABC):
+    """An interface for a backend service that provides access to table data."""
+
+    @abstractmethod
+    def table_schema(self, table_key: TableKey) -> Tuple[pa.Schema, Optional[pa.Schema]]:
+        """ Returns the table data schema and optionally the partitioning column schema for the table with the given
+        table key.
+        The table data schema is not required to include the partitioning columns defined in the partitioning schema.
+
+        Args:
+            table_key (TableKey): the table key
+
+        Returns:
+            Tuple[pa.Schema, Optional[pa.Schema]]: a tuple of the table schema and the optional schema for the
+                partitioning columns
```

Comment:
The "optional" partitioning column schema description in this pydoc is not clear.  Is it optional because it is only present if the table has partitioning columns?  If not, what is making it optional?  When should the user expect to get a value vs None?